torchtext/vocab/vocab_factory.py

```diff
@@ -49,7 +51,12 @@ def vocab(ordered_dict: Dict, min_freq: int = 1,
         ordered_dict.pop(token, None)
 
     tokens = []
-    for token, freq in ordered_dict.items():
+    # Save room for special tokens
+    max_tokens = (max_tokens or float('inf')) - len(specials)
+    for i, (token, freq) in enumerate(ordered_dict.items()):
+        # Save room
+        if i >= max_tokens:
+            break
```

Comment:
One issue I see with this approach is if the `ordered_dict` is not sorted in ascending order or sorted by frequency at all. For example, if the dict is sorted in ascending order, and `max_tokens` equals `10`, it could be possible that the first 10 tokens do not meet the min frequency to be included in the `tokens` list. But because we are using the count from `enumerate()` we would break out of the for-loop even after skipping the first 10 tokens in the `ordered_dict`. In the baseline implementation, we would iterate through all tokens regardless of how they were sorted so this wouldn't be a problem. 

I guess if we assume that the `ordered_dict` will always be sorted by frequency, we can avoid solving for this edge case. @parmeet do you have any thoughts on this?

Reply:
This is a good point. The ordering is enforced when building it using factory functions, but not strictly enforced if someone were to construct their own by calling `vocab` directly. One idea might be to push the sorting to `vocab`, another might be to assert order in `vocab`. I'm interested in your thoughts.

Reply:
Sorting would take `O(n log n)`, but seems like `Counter.most_common` takes `O(n log k)` according to [this thread](https://stackoverflow.com/questions/29240807/python-collections-counter-most-common-complexity). Wonder if it might be more computationally efficient to convert to `Counter` first (assuming the constant time of conversion does not exceed the theoretical gain when k << n).

Reply:
I guess the principle behind vocab factory API is to avoid any assumptions on how user want to construct their vocab since this is the most primitive and closest to the underlying C++ layer. I would be bit cautious of addition more options or  assumptions on the ordering here. For instance, what if there is a pre-existing vocab/order with the corresponding embedding table? We can certainly create higher order factory functions (build_vocab_from_iterator is one example) that can provide more functionality out-of-the-box, and uses vocab factory function to return the actual object.

Reply:
So does it make sense to pull the max_tokens up a layer to the build function and not let that trickle down to vocab? That's an easy enough change to make.

Reply:
yupp, that would be my recommendation, this would help us avoid making assumptions in the primitive factory function :).

Reply:
Awesome -- will fix quickly.